clientconn.go

```diff
@@ -931,184 +926,289 @@ func (ac *addrConn) printf(format string, a ...interface{}) {
-			ridx = 0 // Start connecting from the beginning.
-		} else {
-			// Continue trying to connect with the same deadlines.
-			connectRetryNum = ac.connectRetryNum
-			backoffDeadline = ac.backoffDeadline
-			connectDeadline = ac.connectDeadline
-			ac.backoffDeadline = time.Time{}
-			ac.connectDeadline = time.Time{}
-			ac.connectRetryNum = 0
+		if ac.state == connectivity.Shutdown {
+			ac.mu.Unlock()
+			return
 		}
+		if ac.ready != nil {
+			close(ac.ready)
+			ac.ready = nil
+		}
+		ac.transport = nil
+
+		backoffIdx := ac.backoffIdx
+		backoffFor := ac.dopts.bs.Backoff(backoffIdx)
+
+		// This will be the duration that dial gets to finish.
+		dialDuration := getMinConnectTimeout()
+		if backoffFor > dialDuration {
+			// Give dial more time as we keep failing to connect.
+			dialDuration = backoffFor
+		}
+		start := time.Now()
+		connectDeadline := start.Add(dialDuration)
+		ac.backoffDeadline = start.Add(backoffFor)
+		ac.connectDeadline = connectDeadline
+
+		ac.mu.Unlock()
+
+		ac.cc.mu.RLock()
+		ac.dopts.copts.KeepaliveParams = ac.cc.mkp
+		ac.cc.mu.RUnlock()
+
+		ac.mu.Lock()
+
 		if ac.state == connectivity.Shutdown {
 			ac.mu.Unlock()
-			return errConnClosing
+			return
 		}
+
 		ac.printf("connecting")
 		if ac.state != connectivity.Connecting {
 			ac.updateConnectivityState(connectivity.Connecting)
 			ac.cc.handleSubConnStateChange(ac.acbw, ac.state)
 		}
-		// copy ac.addrs in case of race
-		addrsIter := make([]resolver.Address, len(ac.addrs))
-		copy(addrsIter, ac.addrs)
+
+		addr := ac.addrs[ac.addrIdx]
 		copts := ac.dopts.copts
 		ac.mu.Unlock()
-		connected, err := ac.createTransport(connectRetryNum, ridx, backoffDeadline, connectDeadline, addrsIter, copts, resetBackoff)
-		if err != nil {
-			return err
-		}
-		if connected {
-			return nil
-		}
-	}
-}
 
-// createTransport creates a connection to one of the backends in addrs.
-// It returns true if a connection was established.
-func (ac *addrConn) createTransport(connectRetryNum, ridx int, backoffDeadline, connectDeadline time.Time, addrs []resolver.Address, copts transport.ConnectOptions, resetBackoff chan struct{}) (bool, error) {
-	for i := ridx; i < len(addrs); i++ {
-		addr := addrs[i]
 		if channelz.IsOn() {
 			channelz.AddTraceEvent(ac.channelzID, &channelz.TraceEventDesc{
 				Desc:     fmt.Sprintf("Subchannel picks a new address %q to connect", addr.Addr),
 				Severity: channelz.CtINFO,
 			})
 		}
-		target := transport.TargetInfo{
-			Addr:      addr.Addr,
-			Metadata:  addr.Metadata,
-			Authority: ac.cc.authority,
+
+		if err := ac.createTransport(backoffIdx, addr, copts, connectDeadline); err != nil {
+			// errReadTimeOut indicates that the handshake was not received before
+			// the deadline. We exit here because the transport's reader goroutine will
+			// use onClose to reset the transport.
+			if err == errReadTimedOut {
+				return
+			}
+
+			continue
+		}
+
+		return
+	}
+}
+
+var errReadTimedOut = errors.New("read timed out")
```

Comment:
This is only ever used on line 1020 as a check -- is something broken or do we not need this at all?

Reply:
Done. I think this was an artifact from back when we were using `conn.SetReadDeadline`.